### PR TITLE
feat(rust, python): add dt.quarter_start, dt.quarter_end, dt.year_start and dt.year_end

### LINKED ISSUE #14511: Add dt.quarter_start, dt.quarter_end, dt.year_start and dt.year_end
### Description

I need date_range capabilities similar to pandas to conveniently get week, quarter, and year starts and ends.
I propose to implement quarter_start, quarter_end, year_start and year_end methods from pandas similarly to https://github.com/pola-rs/polars/pull/8435
PR for quarter and year implementation is: #14513
I will add tests and complete documentation once we discuss implementation details.

@mcrumiller and @MarcoGorelli could be interested as you were participating in the #8435

### PR DESCRIPTION
**Pull request by SozinM**
closes #14511

- [x] Implement functionality
- [x] Implement tests
- [x] Correct docs

### CODE REVIEW

File: crates/polars-time/src/quarter_end.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+pub trait PolarsQuarterEnd {
+    fn quarter_end(&self, time_zone: Option<&Tz>) -> PolarsResult<Self>
+    where
+        Self: Sized;
+}
+
+impl PolarsQuarterEnd for DatetimeChunked {
+    fn quarter_end(&self, time_zone: Option<&Tz>) -> PolarsResult<Self> {
```
Comment (mcrumiller) [inline crates/polars-time/src/quarter_end.rs:32]: This function is identical down to the character to `month_end`. Can't we reuse some of the code? If we're going to have month/quarter/year end functions, and they involve computing a date offset and performing the offset, then they should all be reusing the same machinery except for the offset calculation.

Comment (SozinM) [reply crates/polars-time/src/quarter_end.rs:32]: Yes, ofc
It's POC for the function and once We decided that we will implement the in polars (and not move to polars-ext, or straight up decided not to implement) i will perform refactoring

**Comment by MarcoGorelli** [comment]
thanks - looks like it's on the right track, will do a full review next week

File: py-polars/polars/series/datetime.py [hunk 2]
```diff
@@ -1913,6 +1913,131 @@ def month_end(self) -> Series:
...
+
+        Returns
+        -------
+        Series
+            Series of data type :class:`Date` or :class:`Datetime`.
+
+        Notes
+        -----
+        If you're coming from pandas, you can think of this as a vectorised version
+        of `pandas.tseries.offsets.QuarterBegin().rollback(datetime)`.
```
Comment (MarcoGorelli) [inline py-polars/polars/series/datetime.py:1928]: this should be
```
pandas.tseries.offsets.QuarterBegin(startingMonth=1).rollback(datetime)
```

(as an aside, `startingMonth` is a very odd argument name 😄 )

File: py-polars/polars/expr/datetime.py [hunk 3]
```diff
@@ -2162,6 +2162,192 @@ def month_end(self) -> Expr:
...
+        ...     {
+        ...         "dates": pl.datetime_range(
+        ...             datetime(2000, 1, 15, 2),
+        ...             datetime(2002, 12, 15, 2),
+        ...             "6mo",
+        ...             eager=True,
+        ...         )
+        ...     }
+        ... )
+        >>> df.select(pl.col("dates").dt.year_end())
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:2334]: can we use `with_columns` for these examples? I realise `month_end` already uses `select` - do you want to change that to `with_columns` first in a precursor PR?

File: py-polars/docs/source/reference/series/temporal.rst [hunk 4]
```diff
@@ -38,6 +38,10 @@ The following methods are available under the `Series.dt` attribute.
     Series.dt.month
     Series.dt.month_end
     Series.dt.month_start
+    Series.dt.quarter_end
+    Series.dt.quarter_start
+    Series.dt.year_end
+    Series.dt.year_start
```
Comment (MarcoGorelli) [inline py-polars/docs/source/reference/series/temporal.rst:44]: shall we keep the alphabetical ordering?

File: crates/polars-time/src/lib.rs [hunk 5]
```diff
@@ -1,11 +1,11 @@
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
 mod base_utc_offset;
 pub mod chunkedarray;
+mod date_end;
 mod date_range;
+mod date_start;
```
Comment (MarcoGorelli) [inline crates/polars-time/src/lib.rs:6]: 🤔 perhaps `period_start` / `period_end`?

File: crates/polars-plan/src/dsl/dt.rs [hunk 6]
```diff
@@ -205,6 +205,34 @@ impl DateLikeNameSpace {
...
+    }
+
+    /// Roll forward to the last day of the month.
+    #[cfg(feature = "date_offset")]
+    pub fn quarter_end(self) -> Expr {
+        self.0
+            .map_private(FunctionExpr::TemporalExpr(TemporalFunction::QuarterEnd))
+    }
+
+    /// Roll backward to the first day of the month.
```
Comment (MarcoGorelli) [inline crates/polars-plan/src/dsl/dt.rs:222]: copy-pasta comment 😉

**Review by MarcoGorelli — COMMENTED** [review]
Nice work!

File: py-polars/polars/expr/datetime.py (hunk 3 above)
Comment (SozinM) [reply py-polars/polars/expr/datetime.py]: I added it to this PR. 
Do you specifically want to split it off?

**Comment by SozinM** [comment]
All comments were addressed @MarcoGorelli

File: crates/polars-plan/src/dsl/function_expr/datetime.rs [hunk 7]
```diff
@@ -105,6 +113,14 @@ impl TemporalFunction {
             MonthStart => mapper.with_same_dtype(),
             #[cfg(feature = "date_offset")]
             MonthEnd => mapper.with_same_dtype(),
+            #[cfg(feature = "date_offset")]
```
Comment (mcrumiller) [inline crates/polars-plan/src/dsl/function_expr/datetime.rs:116]: FYI for brevity you can split this with the `|` symbol:

```rust
#[cfg(feature = "date_offset")]
MonthStart | MonthEnd | QuarterStart | QuarterEnd | YearStart | YearEnd => mapper.with_same_dtype(),
```

Comment (SozinM) [reply crates/polars-plan/src/dsl/function_expr/datetime.rs:116]: nice one, thank you!

**Comment by MarcoGorelli** [comment]
thanks @SozinM for updating - could you update the merge conflict please? It should just be a matter of passing `NonExistent::Raise` to `replace_time_zone`

File: crates/polars-time/src/period_start.rs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+pub(crate) enum RollUnit {
+    Month,
+    Quarter,
+    Year,
+}
+
+#[cfg(feature = "timezones")]
+use crate::utils::{try_localize_datetime, unlocalize_datetime};
+
+// roll backward to the first day of the month
```
Comment (MarcoGorelli) [inline crates/polars-time/src/period_start.rs:20]: this comment needs updating

File: crates/polars-time/src/period_start.rs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        #[cfg(feature = "timezones")]
+        Some(tz) => unlocalize_datetime(timestamp_to_datetime(t), tz),
+        _ => timestamp_to_datetime(t),
+    };
+    let date = match roll_unit {
+        RollUnit::Month => NaiveDate::from_ymd_opt(ts.year(), ts.month(), 1).ok_or_else(|| {
+            polars_err!(
+                ComputeError: format!("Could not construct date {}-{}-1", ts.year(), ts.month())
+            )
+        })?,
```
Comment (MarcoGorelli) [inline crates/polars-time/src/period_start.rs:38]: codecov is showing these errors an uncovered

I don't really how they could error anyway, as we're constructing naive dates where the `date` component is `1`. How about just using `.expect("we know we're creating a valid date")`?

File: py-polars/polars/expr/datetime.py [hunk 10]
```diff
@@ -2142,35 +2142,221 @@ def month_end(self) -> Expr:
...
+
+        Returns
+        -------
+        Expr
+            Expression of data type :class:`Date` or :class:`Datetime`.
+
+        Notes
+        -----
+        If you're coming from pandas, you can think of this as a vectorised version
+        of `pandas.tseries.offsets.QuarterEnd().rollforward(datetime)`.
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/datetime.py:2235]: lol wtf pandas QuarterEnd starts in January but QuarterBegin starts in December? 🤯 

(I'm not asking you to do anything here, I'm just...amazed)

File: crates/polars-time/src/period_end.rs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+use polars_core::utils::arrow::temporal_conversions::{
+    timestamp_ms_to_datetime, MILLISECONDS, SECONDS_IN_DAY,
+};
+
+use crate::period_start::{
+    get_timestamp_closures, roll_month_backward, roll_quarter_backward, roll_year_backward,
+};
+use crate::windows::duration::Duration;
+
+// roll forward to the last day of the month
```
Comment (MarcoGorelli) [inline crates/polars-time/src/period_end.rs:13]: can we either copy this comment to the respective functions below, or remove it?

**Review by MarcoGorelli — COMMENTED** [review]
Nice one @SozinM ! this looks close

Just a couple of minor comments and it should be good to go

**Comment by SozinM** [comment]
@MarcoGorelli done :)

**Review by MarcoGorelli — APPROVED** [review]
Nice one, thanks @SozinM !

**Comment by SozinM** [comment]
@alexander-beedie Hi! Would you like to review this?

**Comment by SozinM** [comment]
@MarcoGorelli could you suggest a person to ask for an additional review?

**Comment by alexander-beedie** [comment]
> @alexander-beedie Hi! Would you like to review this? 

This is one for @MarcoGorelli 😎👍
But the functionality looks useful!

**Comment by SozinM** [comment]
@alexander-beedie Hi!
Rebased, take a look please :)

**Comment by SozinM** [comment]
Hi @reswqa , @orlp !
Maybe you will save the day? (or save this PR from being buried in history)

**Comment by SozinM** [comment]
Hi @ritchie46 maybe you will be able to help?

**Comment by MarcoGorelli** [comment]
hi @SozinM - seeing as this is still open, it may be good to get your thoughts on https://github.com/pola-rs/polars/issues/16094

if `month_end` were indeed disallowed for `Datetime`s, it may make sense to remove it from this PR too. do you have thoughts on that?

**Comment by stinodego** [comment]
I've felt a bit apprehensive pressing the merge button, because I'm not sure these methods are really that useful. Don't we already offer all the right tools to do this yourself? E.g. 

```python
from datetime import date

import polars as pl

df = pl.DataFrame({"date": [date(2024, 2, 29)]})
result = df.with_columns(
    year_end=pl.date(year=pl.col("date").dt.year(), month=12, day=31)
)
print(result)
```
```
shape: (1, 2)
┌────────────┬────────────┐
│ date       ┆ year_end   │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2024-02-29 ┆ 2024-12-31 │
└────────────┴────────────┘
```

The API bloat just doesn't seem worth it.

**Comment by SozinM** [comment]
Okay, this PR got a little out of hand so I will close it. If somebody wants to pick it up - feel free to repurpose the code :)

**Pull request closed without merging**